### DISCUSSION TITLE
Planung weiterer Versionen

### DISCUSSION THREAD

**Discussion by schillermann** [schillermann/trolleydienst-php#34] [Ideas]
Hallo Zusammen,

hier wollte ich mit dem Team die Möglichkeiten der Software Erweiterung und Planung besprechen.

## Welche Möglichkeiten gibt es?
1. Redesign der Software 
2. UI Logic in den Browser auslagern und die Business Logic per API zugänglich machen
3. Modul System integrieren

Welche Varianten haltet ihr für besonders wichtig? Und welche sind unnötig? Habt ihr noch weitere Ideen, welche hier auf der Liste fehlen?

**Comment by schillermann** [comment]
@ShahinS2, @mikecopestake, @user-kh und @S-Pfeil bin gespannt auf eure Meinungen.
Hoffe ich habe da keinen aus dem Team vergessen.

**Comment by schillermann** [comment]
Wenn die UI Logic in den Browser als WebApp ausgelagert werden soll, muss das vor dem Umbau auf Module passieren. In der WebApp müssen Module anders aufgebaut werden als wenn die Seiten Server seitig gebaut werden.

**Comment by schillermann** [comment]
Im Projekt verwenden wir anscheinend immer mehr JavaScript für die UI oder planen es für die Zukunft. Ganz darauf verzichtet können wir wahrscheinlich nicht. Dann wäre meine nächste Frage ob es nicht ein logischer nächster Schritt wäre bei der UI ganz auf Javascript zu setzen. Dann hat man hier kein Mischmasch aus PHP und Javascript mehr. Was denkt ihr?

**Comment by user-kh** [comment]
Hallo an alle!

Ich stehe hinter dem Vorschlag von @schillermann.
Meiner Meinung nach sind alle drei Punkte wichtig und sinnvoll. Ich würde Sie nur etwas konkretisieren:

1. Redesign der Software
   a: Benutzerdefinierte Farben; Optik kann die eigenen Bedürfnisse angepasst werden 
   b: Neueres frisches Design

2. UI Logic in den Browser auslagern und die Business Logic per API zugänglich machen
    a: PHP Backend mit Login-Logik
    b: APIs für Login und Datenbank zur Anbindung an die Module

3. Modul System integrieren
    a: JS Web Components (ggf. mit Framework); alternativ mit fest definierten Klassen und *guter Dokumentation* für externe Entwickler der Module

Was meint hier?
LG

**Comment by schillermann** [comment]
Um eine bessere Vorstellung von einer SPA mit Web Components bekommen zu können habe ich mal eine [Beispiel Anwendung](https://github.com/schillermann/web-components-spa-example) erstellt.
Fragt sich nur ob andere Entwickler mit dem Aufbau umgehen können, oder ob es nicht zu kompliziert ist um es verstehen zu können.

@user-kh das ist auch eine interessante Idee mit einem austauschbaren Theme.
Wenn man ein eigenes Theme erstellen möchte müsste man alle Dateien im Ordner `/public/layout` und `/public/page` duplizieren und anpassen.

**Comment by schillermann** [comment]
Hab die Html Templates der [Beispiel Anwendung](https://github.com/schillermann/web-components-spa-example) in ein [Default Theme](https://github.com/schillermann/web-components-spa-example/tree/main/public/themes/default) ausgelagert.
Wie findet ihr die Umsetzung?

**Comment by schillermann** [comment]
Hab die [Ordner Struktur angepasst](https://github.com/schillermann/web-components-spa-example/commit/cb94bdb4eb55e0e15e92e653f55bf57c9d2dec00), für eine bessere Übersicht der Web Components.
Hoffe das die neue Struktur für euch etwas verständlicher ist.

Als nächstes kommt der Modul Aufbau dran, um sie in die Navigation einzubinden.

**Comment by S-Pfeil** [comment]
Ich kann da leider nicht mitreden was da besser ist usw. Hab leider kein Ahnung davon.
Was aber gut wäre, ist das die [Demo](https://trolleydienst-demo.schillermann.de/) funktioniert denn das tut sie seit Wochen nicht. Ich bekomme da immer wieder nur einen weißen Bildschirm. 
Danke für Eure Arbeit!
Lg S.Pfeil

**Comment by schillermann** [comment]
Hi @S-Pfeil,

danke dir für den Hinweis mit der [Demo](https://trolleydienst-demo.schillermann.de/) Version.
Habe sie jetzt auf den neusten Stand gebracht.

**Comment by schillermann** [comment]
@user-kh in der Beispiel Anwendung werden nun in der Navigation die Module ausgegeben.
So sieht eine [Config Datei](https://github.com/schillermann/web-components-spa-example/blob/main/public/modules/admin/config.json) in einem Modul aus. 
Wie findest du den Aufbau bis jetzt?

**Comment by schillermann** [comment]
Die Kern Funktionalität habe ich in das [Modul Planner](https://github.com/schillermann/web-components-spa-example/tree/main/public/modules/planner) ausgelagert.
Somit ist alles ein Modul.
Ich bin mir noch nicht sicher ob ich das Profil und die Benutzerverwaltung vom [Admin Modul](https://github.com/schillermann/web-components-spa-example/tree/main/public/modules/admin) nicht auch im [Modul Planner](https://github.com/schillermann/web-components-spa-example/tree/main/public/modules/planner) umziehen sollte.
Jedes Modul bringt seine eigenen Html und Css Dateien mit.
Damit sind Themes nicht mehr möglich.
Außer ihr habt noch eine andere Idee wie man Themes nutzen kann.

Das Projekt ist zum testen auf [planner.schillermann.de](https://planner.schillermann.de) verfügbar.
Mit Benutzer admin und Passwort admin könnt ihr euch einloggen.
Freue mich auf Anregungen.

**Reply by user-kh** [reply]
Hallo an alle!

Ich bin derzeit noch in der Entwicklung der Versionen 1.xx.xx.
Ich gehe davon aus, dass wir dieses Jahr noch ein größeres Release rausbringen.

Sobald die neuen Features integriert sind, kann ich mich mehr dem Projekt 2.0 widmen.

Ich würde das Profil und die Benutzerverwaltung im Admin Modul belassen. Die Daten müssen ja modulübergeifend zur Verfügung stellen.
Damit hat man eine gemeinsame Datenbasis.

Ich denke vom Sicherheitsaspekt ist es auch besser.

Die Demo auf planner.schillermann.de scheint nicht (mehr) zu funktionieren!?

LG

**Comment by schillermann** [comment]
Hallo Zusammen,

den [planner.schillermann.de](https://planner.schillermann.de/) habe ich aktualisiert.
Man kann sich mit dem Username admin und dem Passwort admin anmelden, welcher in der Datenbank hinterlegt ist.
Als nächstes werden ich mich dem Design der Benutzerliste widmen.

**Comment by user-kh** [comment]
Super die Demo läuft 😉👍

**Comment by schillermann** [comment]
Den Style der Benutzerliste habe ich angepasst.
Später können die Zellen in der mobilen Ansicht untereinander dargestellt werden.
Wenn man auf einen der Benutzer klickt soll sich später die Seite mit den Benutzer Details öffnen. 
Reicht euch die Anzahl der Spalten?

**Comment by schillermann** [comment]
Guten Abend Zusammen,
hab ein Benutzer Formular erstellt.
Ihr könnt es sehen, wenn ihr in der Users List auf den Admin klickt.
Das Ändern der Daten folgt noch.
